crates/external_services/src/grpc_client.rs

```diff
@@ -0,0 +1,238 @@
+        response: Vec<RoutableConnectorChoiceWithStatus>,
+    ) -> DRResult<UpdateSuccessRateWindowResponse>;
+}
+
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum DRError {
+    #[error("Error buling the gRPC Client for communication")]
+    ClientBuildingFailed,
+    #[error("Error getting a response from the gRPC Server")]
+    GrpcServerResponseFailure,
+}
+/// Struct that contains the settings required to construct an Grpc client.
+#[derive(Debug, Clone, serde::Deserialize, serde::Serialize, Default)]
+pub struct GrpcClientSettings {
+    pub dynamic_routing_client: DynamicRoutingClientConfig,
+}
+#[derive(Debug, Clone, serde::Deserialize, serde::Serialize, Default)]
+pub struct DynamicRoutingClientConfig {
+    pub host: String,
+    pub port: u16,
+    pub enabled: bool,
+}
+
+impl GrpcClientSettings {
+    /// # Panics
+    ///
+    /// This function will panic if it fails to establish a connection with the gRPC server.
+    #[allow(clippy::expect_used)]
+    pub async fn get_grpc_client_interface(&self) -> GrpcClients {
+        let grpc_connect = self
+            .dynamic_routing_client
+            .clone()
+            .get_dynamic_routing_connection()
+            .await
+            .expect("Failed to establish a connection with the gRPC Server");
+        logger::debug!("Connection established with Grpc Server");
+        GrpcClients {
+            dynamic_routing: grpc_connect,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct GrpcClients {
+    pub dynamic_routing: RoutingStrategy,
+}
+
+#[derive(Debug, Clone)]
+pub struct RoutingStrategy {
+    pub success_rate_client: Option<SuccessRateCalculatorClient<Channel>>,
+}
+
+impl DynamicRoutingClientConfig {
+    pub async fn get_dynamic_routing_connection(
+        self,
+    ) -> Result<RoutingStrategy, Box<dyn std::error::Error>> {
+        let success_rate_client = if self.enabled {
+            let uri = format!("http://{}:{}", self.host, self.port);
+            let channel = tonic::transport::Endpoint::new(uri)?.connect().await?;
+            Some(SuccessRateCalculatorClient::new(channel))
+        } else {
+            None
+        };
+        Ok(RoutingStrategy {
+            success_rate_client,
+        })
+    }
+}
+#[async_trait::async_trait]
+impl SuccessBasedDynamicRouting for SuccessRateCalculatorClient<Channel> {
+    async fn calculate_success_rate(
+        &self,
+        id: id_type::ProfileId,
+        dynamic_routing_config: DynamicRoutingConfig,
+        label_input: Vec<RoutableConnectorChoice>,
+    ) -> DRResult<CalSuccessRateResponse> {
+        let params = dynamic_routing_config
+            .params
+            .map(|vec| {
+                if !vec.is_empty() {
+                    vec.into_iter()
+                        .map(|param| param.to_string())
+                        .collect::<Vec<String>>()
+                        .join(":")
+                } else {
+                    String::default()
+                }
```

Comment:
```suggestion
                input.into_iter().fold(String::new(), |mut acc, cur| {
                        if !acc.is_empty() {
                            acc.push(':');
                        }
                        acc.push_str(cur);
                        acc
                });
```

This can solve for the following:
1. Less memory allocations
2. The outer check for empty vector
3. Default behavior is implicit